retrieval/discovery/consul/consul.go

```diff
@@ -0,0 +1,259 @@
+	"golang.org/x/net/context"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+const (
+	watchTimeout  = 30 * time.Second
+	retryInterval = 15 * time.Second
+
+	// addressLabel is the name for the label containing a target's address.
+	addressLabel = model.MetaLabelPrefix + "consul_address"
+	// nodeLabel is the name for the label containing a target's node name.
+	nodeLabel = model.MetaLabelPrefix + "consul_node"
+	// tagsLabel is the name of the label containing the tags assigned to the target.
+	tagsLabel = model.MetaLabelPrefix + "consul_tags"
+	// serviceLabel is the name of the label containing the service name.
+	serviceLabel = model.MetaLabelPrefix + "consul_service"
+	// serviceAddressLabel is the name of the label containing the (optional) service address.
+	serviceAddressLabel = model.MetaLabelPrefix + "consul_service_address"
+	//servicePortLabel is the name of the label containing the service port.
+	servicePortLabel = model.MetaLabelPrefix + "consul_service_port"
+	// datacenterLabel is the name of the label containing the datacenter ID.
+	datacenterLabel = model.MetaLabelPrefix + "consul_dc"
+	// serviceIDLabel is the name of the label containing the service ID.
+	serviceIDLabel = model.MetaLabelPrefix + "consul_service_id"
+)
+
+// Discovery retrieves target information from a Consul server
+// and updates them via watches.
+type Discovery struct {
+	client           *consul.Client
+	clientConf       *consul.Config
+	clientDatacenter string
+	tagSeparator     string
+	watchedServices  []string // Set of services which will be discovered.
+}
+
+// NewDiscovery returns a new Discovery for the given config.
+func NewDiscovery(conf *config.ConsulSDConfig) (*Discovery, error) {
+	clientConf := &consul.Config{
+		Address:    conf.Server,
+		Scheme:     conf.Scheme,
+		Datacenter: conf.Datacenter,
+		Token:      conf.Token,
+		HttpAuth: &consul.HttpBasicAuth{
+			Username: conf.Username,
+			Password: conf.Password,
+		},
+	}
+	client, err := consul.NewClient(clientConf)
+	if err != nil {
+		return nil, err
+	}
+	cd := &Discovery{
+		client:          client,
+		clientConf:      clientConf,
+		tagSeparator:    conf.TagSeparator,
+		watchedServices: conf.Services,
+	}
+	// If the datacenter isn't set in the clientConf, let's get it from the local Consul agent
+	// (Consul default is to use local node's datacenter if one isn't given for a query).
+	if clientConf.Datacenter == "" {
+		info, err := client.Agent().Self()
+		if err != nil {
+			return nil, err
+		}
+		cd.clientDatacenter = info["Config"]["Datacenter"].(string)
+	} else {
+		cd.clientDatacenter = clientConf.Datacenter
+	}
+	return cd, nil
+}
+
+// shouldWatch returns whether the service of the given name should be watched.
+func (cd *Discovery) shouldWatch(name string) bool {
+	// If there's no fixed set of watched services, we watch everything.
+	if len(cd.watchedServices) == 0 {
+		return true
+	}
+	for _, sn := range cd.watchedServices {
```

Comment:
Yes, it will be O(n) but in this use case it will plenty faster than using a hash map. It really doesn't matter here, but we don't have to build a hash map to begin with, which makes the code more self-explanatory to me.

n will typically be around 1-10.
Note that these are the services we explicitly configured to watch.

Let's look into metrics in subsequent PRs.